Update rubber band after editing nodes

dgoedkoop opened the pull request:
It is a nice new functionality that the node editor puts a rubber band on the feature currently being edited. However, so far this rubber band is not updated when the nodes of the feature are actually edited. This PR is to ~~fix~~ change that.

It uses a geometryChanged signal rather than just adding updateSelectFeature() in a few places, because only in this way the rubber band is also correctly updated after an undo operation. It splits updateSelectFeature() into two parts, because the geometryChanged signal is emitted with the new (or in case of an undo: previous) geometry as a parameter, before the geometry of the feature is actually changed.

Edit: As result of the discussion, an option has been added to choose between both kinds of behaviour.

dgoedkoop committed:
Update rubber band after editing nodes

It has to be done this way. Just adding updateSelectFeature() in a few places
is not enough, because then the rubber band will still not be updated after
an undo operation.

m-kuhn commented on src/app/nodetool/qgsmaptoolnodetool.cpp:
```diff
@@ -114,6 +114,11 @@ void QgsMapToolNodeTool::canvasMoveEvent( QgsMapMouseEvent* e )
   {
     if ( mMoveRubberBands.empty() )
     {
+      if ( mSelectRubberBand )
```
needless to check, just delete and set to 0. If it's alread 0 nothing will happen.

dgoedkoop replied:
Should this be changed in cleanTool() too, then?

m-kuhn replied:
Yes :)

dgoedkoop committed:
Remove some unnecessary null checks in node tool

dgoedkoop commented:
Wait, this is not ok yet. It makes qgis crash if you delete the very last vertex of a geometry using the node editor.

m-kuhn commented:
No worries, you can leave it open, so people can review it anyway.

gioman commented:
I think that non updating the rubber band was on purpose: this is how other gis packages work (most noticeably AG) and is what people expect because it is useful to keep track on how the line/boundary was before moving node/s. So from a user point of view is a -1.

dgoedkoop committed:
Prevent crash when deleting last node of geometry using node editor

dgoedkoop commented:
@gioman: I didn't think of that but it makes sense. And I didn't know it before, because at my work we only use qgis.

dgoedkoop commented:
ArcGis' behaviour also seems to be similar. So I think this PR can be closed/cancelled.

m-kuhn commented:
Thank you @dgoedkoop and @gioman

nyalldawson commented:
@gioman @m-kuhn @dgoedkoop Can we start this discussion again? After using the new behaviour for a couple of weeks, I find it a HUGE step back from editing in older QGIS versions. The "ghost" of the old feature just clutters up the display with no added benefit for me, and I've resorted to a make-one-change-click-save-immediately-to-hide-old-geometry-ghost behaviour.

@gioman if you find this new behaviour better, can we make it a (non-default) option please?

gioman commented:
@nyalldawson Hi Nyall! I don't do much digitizing, in this case I just left here the opinion of many users that are switching from proprietary, or in the process to: they often say to me that seeing how it was the geometry before moving nodes (and saving) it is a good feature (and personally I like it too). But of course if the general opinion here is different I have nothing against "fixing" it or making it optional. Cheers!

dgoedkoop commented:
If it should be an option, then what should be the exact behaviour when disabled? Should the rubber band then follow all changes, as this PR would do? Or should it not be shown at all (e.g. no change from 2.12 if disabled), to avoid confusion between the rubber band as selected feature marker and the same rubber band as 'ghost'?

nyalldawson commented:
I think "follow all changes" would be best. Then you'd still be able to identify which geometry you are editing if there's overlapping features

dgoedkoop committed:
Made the behaviour configurable and remove conflicting change

In the options dialog, you can now select which behaviour is
desired: the rubber band als 'ghost' of the line before node
editing, or the rubber band as dynamically updated selection
marker (default).

dgoedkoop commented:
PR #2490 (merged two days ago) tried to do the same as this PR, in commit 7bdaf53a4b012e7525a6a1d5036ff530c2748be8. However, that is not the way to do it, because it doesn't react to changes using the keyboard (e.g. delete vertex) or to undo/redo operations.

I've updated this PR to replace that change and to add an option in the options dialog.

dgoedkoop committed:
Coding style fixes

dgoedkoop committed:
Make sure default setting is consistent

dgoedkoop commented:
Sorry if my previous comment was a bit harsh. Anyways, @m-kuhn, is this still "Requires Changes" now?

m-kuhn commented:
No worries, I don't think somebody felt offended :)
requires changes was because of your own comment about the crash, happy to change it.

m-kuhn commented:
@3nids if you don't feel responsible, please reassign at will.

3nids commented:
merged in b33b9351f75eaafb23f2a2909f931a07dab19524
